documentation/porting_guidelines.md

```diff
@@ -0,0 +1,400 @@
+`crypto_ops` structure and write the low level driver that communicates with the
+device. Our [crypto.md] file describes how to add and implement a new `struct
+crypto_ops`. Since the communication with crypto blocks tends to be quite
+different depending on what kind of crypto block you have, we haven't written
+how that should be done. It might be that we do that in the future when get hold
+of a device where we can use the crypto block.
+
+## 7. Power Management / PSCI
+In section 2 when we talked about the file `main.c`, we added a couple of
+handlers related to power management, we are talking about the following lines:
+```
+	.cpu_on = cpu_on_handler,
+	.cpu_off = pm_do_nothing,
+	.cpu_suspend = pm_do_nothing,
+	.cpu_resume = pm_do_nothing,
+	.system_off = pm_do_nothing,
+	.system_reset = pm_do_nothing,
+```
+The only function that actually does something there is the `cpu_on` function,
+the rest of them are stubbed. The main reason for that is because we think that
+how to suspend and resume is a device dependant thing. The code in OP-TEE is
+prepared so that callbacks etc from ARM Trusted Firmware will be routed to
+OP-TEE, but since the function(s) are just stubbed we won't do anything and just
+return. In a real production device, you would probably want to save and restore
+CPU states, secure hardware IPs' registers and TZASC and other memory firewall
+related setting when these callbacks are being called.
+
+## 8. Memory firewalls / TZASC
+ARM have defined a system IP / SoC peripheral called TrustZone Adress Space
+Controller (TZASC, see [TZC-380] and [TZC-400]). TZASC can be used configured
+DDR memory into separate regions in the address space, where each region can
+have an individual security level setting. After enabling TZASC, it will perform
+security checks on transactions to memory or peripherals. It is not always the
+case that TZASC is on a device, in some cases the SoC has developed something
+equivalent. In OP-TEE this is very well reflected, i.e., different platforms
+have different ways of protecting its memory. On ARMv8-A platforms we're in most
+of the cases using ARM Trusted Firmware as the boot firmware and there they are
+the ones that configures secure vs non-secure memory using TZASC (see
+[plat_arm_security_setup] in ARM-TF). The takeaway here is that you make sure
+that you have configured whatever memory firewall your device has such that it
+has a secure and a non-secure memory area.
+
+## 9. Trusted Application private/public keypair
+By default all Trusted Applications (TA's) are signed with the pre-generated
+2048-bit RSA development key (private key). This key is located in the `keys`
+folder (in the root of optee_os.git) and is named `default_ta.pem`. This key
+**must** be replaced with your own key and you should **never ever** check-in
+this private key in the source code tree when in use in a real product. The
+recommended way to store private keys are to use some kind of [HSM] (Harware
```

Comment:
or you could change to `The recommended ways to store private keys are to use some kind of ...
`.